### PR TITLE
Remove completion value.

### PR DESCRIPTION
**Pull request by zherczeg**
Remove ecma_completion_value_t, and an extra bit to ecma_value_t to represent errors. From the long list of completion types only normal and error remained.

### CODE REVIEW

**Comment by bzsolt** [comment]
RPi2 performance measurements
master https://github.com/Samsung/jerryscript/commit/977bfa5d2cc5c60eec587e8f406715f415a3c57b

                         Benchmark |              Perf<br>(+ is better)
                        3d-cube.js |  2.686s ->   2.644s :  +1.564%  (+-0.589%) : [+]
            access-binary-trees.js |  1.422s ->   1.408s :  +0.985%  (+-1.705%) : [≈]
                access-fannkuch.js |  8.026s ->   7.954s :  +0.897%  (+-1.187%) : [≈]
                   access-nbody.js |  3.158s ->   3.098s :  +1.900%  (+-1.001%) : [+]
       bitops-3bit-bits-in-byte.js |  1.784s ->   1.758s :  +1.457%  (+-0.809%) : [+]
            bitops-bits-in-byte.js |  2.748s ->   2.718s :  +1.092%  (+-1.029%) : [+]
             bitops-bitwise-and.js |  3.788s ->   3.718s :  +1.848%  (+-1.067%) : [+]
          controlflow-recursive.js |  0.946s ->   0.896s :  +5.285%  (+-1.642%) : [+]
                     crypto-aes.js |  4.640s ->   4.658s :  -0.388%  (+-0.965%) : [≈]
                     crypto-md5.js | 21.582s ->  21.884s :  -1.399%  (+-0.493%) : [-]
                    crypto-sha1.js |  9.764s ->   9.872s :  -1.106%  (+-0.367%) : [-]
              date-format-tofte.js |  2.198s ->   2.186s :  +0.546%  (+-1.142%) : [≈]
              date-format-xparb.js |  1.074s ->   1.066s :  +0.745%  (+-1.480%) : [≈]
                    math-cordic.js |  3.098s ->   2.994s :  +3.357%  (+-0.464%) : [+]
              math-partial-sums.js |  1.698s ->   1.680s :  +1.060%  (+-0.537%) : [+]
             math-spectral-norm.js |  1.536s ->   1.506s :  +1.953%  (+-1.028%) : [+]
                   string-fasta.js |  4.556s ->   4.462s :  +2.063%  (+-0.444%) : [+]
                   Geometric mean: |                Speed up: 1.298% (+-0.252%) : [+]

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 1]
```diff
@@ -199,17 +215,29 @@ ecma_is_value_string (ecma_value_t value) /**< ecma-value */
...
 /**
+ * Check if the value is an error value.
+ *
+ * @return true - if the value contains an error value,
+ *         false - otherwise.
+ */
+bool __attr_pure___ __attr_always_inline___
+ecma_is_value_error (ecma_value_t value) /**< ecma value */
+{
+  return (value & (1 << ECMA_VALUE_ERROR_POS)) != 0;
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-value.c]: I think this should be `return (value & (1ull << ECMA_VALUE_ERROR_POS)) != 0;` like in other places in our code.

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 2]
```diff
@@ -333,7 +379,24 @@ ecma_get_object_from_value (ecma_value_t value) /**< ecma-value */
...
+ * Get the value from an error ecma value
+ *
+ * @return ecma value
+ */
+ecma_value_t __attr_pure___
+ecma_get_value_from_error_value (ecma_value_t value) /**< ecma value */
+{
+  JERRY_ASSERT (ecma_is_value_error (value));
+
+  value = (ecma_value_t) (value & ~(1u << ECMA_VALUE_ERROR_POS));
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-value.c]: We should use a common style for shifting operations (we are mixing 1, 1u, 1ull)

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 3]
```diff
@@ -291,12 +319,30 @@ ecma_make_object_value (const ecma_object_t *object_p) /**< object to reference
...
+ecma_value_t __attr_const___
+ecma_make_error_value (ecma_value_t value) /**< original ecma value */
+{
+  /* Error values cannot be converted. */
+  JERRY_ASSERT (!ecma_is_value_error (value));
+
+  return value | (1u << ECMA_VALUE_ERROR_POS);
+} /* ecma_make_error_value */
+
+ecma_value_t __attr_const___
```
Comment (galpeter) [inline jerry-core/ecma/base/ecma-helpers-value.c]: Missing method comments :)

**Comment by LaszloLango** [comment]
Patch is updated and rebased to master.

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 4]
```diff
@@ -149,20 +149,36 @@ ecma_is_value_boolean (ecma_value_t value) /**< ecma-value */
...
 {
   return (ecma_get_value_type_field (value) == ECMA_TYPE_SIMPLE
           && ecma_get_value_value_field (value) == ECMA_SIMPLE_VALUE_TRUE);
 } /* ecma_is_value_true */
 
 /**
+ * Check if the value is false.
+ *
+ * Warning:
+ *         value must be boolean
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-value.c]: Why this warning? If `value` is not a boolean, does it cause any trouble (assertion, etc.)? I get from the code that the function will simply return `false`, wont it?

Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-value.c]: To tell the truth I just copied&pasted the description from above (see: `ecma_is_value_true`). We can remove that warning message, it looks useless to me too.

File: jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c [hunk 5]
```diff
@@ -76,20 +76,20 @@ ecma_builtin_function_prototype_object_to_string (ecma_value_t this_arg) /**< th
...
                                               ecma_value_t arg2) /**< second argument */
 {
-  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+  ecma_value_t ret_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
 
   /* 1. */
   if (!ecma_op_is_callable (this_arg))
   {
-    ret_value = ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_TYPE));
+    ret_value = ecma_raise_type_error ("");
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c]: This seems to be a bit vague. Shouldn't we report "Incompatible type" at least?

File: jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c [hunk 6]
```diff
@@ -105,7 +105,7 @@ ecma_builtin_function_prototype_object_apply (ecma_value_t this_arg, /**< this a
       /* 3. */
       if (!ecma_is_value_object (arg2))
       {
-        ret_value = ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_TYPE));
+        ret_value = ecma_raise_type_error ("");
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c]: ditto

File: jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c [hunk 7]
```diff
@@ -178,17 +178,17 @@ ecma_builtin_function_prototype_object_apply (ecma_value_t this_arg, /**< this a
...
-static ecma_completion_value_t
+static ecma_value_t
 ecma_builtin_function_prototype_object_call (ecma_value_t this_arg, /**< this argument */
                                              const ecma_value_t *arguments_list_p, /**< list of arguments */
                                              ecma_length_t arguments_number) /**< number of arguments */
 {
   if (!ecma_op_is_callable (this_arg))
   {
-    return ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_TYPE));
+    return ecma_raise_type_error ("");
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c]: ditto

File: jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c [hunk 8]
```diff
@@ -218,20 +218,20 @@ ecma_builtin_function_prototype_object_call (ecma_value_t this_arg, /**< this ar
...
                                              ecma_length_t arguments_number) /**< number of arguments */
 {
-  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+  ecma_value_t ret_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
 
   /* 2. */
   if (!ecma_op_is_callable (this_arg))
   {
-    ret_value = ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_TYPE));
+    ret_value = ecma_raise_type_error ("");
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c]: ditto

File: jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c [hunk 9]
```diff
@@ -376,29 +370,29 @@ ecma_builtin_function_prototype_object_bind (ecma_value_t this_arg, /**< this ar
...
  */
-ecma_completion_value_t
+ecma_value_t
 ecma_builtin_function_prototype_dispatch_construct (const ecma_value_t *arguments_list_p, /**< arguments list */
                                                     ecma_length_t arguments_list_len) /**< number of arguments */
 {
   JERRY_ASSERT (arguments_list_len == 0 || arguments_list_p != NULL);
 
-  return ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_TYPE));
+  return ecma_raise_type_error ("");
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-function-prototype.c]: ditto (I stop ditto'ing now, wont comment if there are more of these)

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.c [hunk 10]
```diff
@@ -804,7 +804,7 @@ ecma_builtin_global_object_decode_uri_helper (ecma_value_t uri __attr_unused___,
 
     if (!lit_read_code_point_from_hex (input_char_p + 1, 2, &decoded_byte))
     {
-      ret_value = ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_URI));
+      ret_value = ecma_raise_uri_error ("");
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.c]: Same with this type of error. But there is no previous use case in the code base, so I'm not sure what to report.
(Several similar occurences below.)

**Comment by akiss77** [comment]
Made my observations above. (Those idiomatically used empty string error messages smell bad. However, it might need all-over-the-code changes to remove them properly, e.g., with functions that don't even expect a string, if they are not used whatsoever.) The rest is OK, as far as I can tell.

**Comment by LaszloLango** [comment]
@akiss77, maintaining and improving the error messages would be great and we need to do that. Do you mind if I raise an issue for that and do it in a separate PR?

**Comment by akiss77** [comment]
@LaszloLango, the patch is good with me as it is. And we don't necessarily have to come up with fancy error messages if we don't want to. But then we should declare that and then we can remove the string argument of those `ecma_raise_*` functions. The use of `""` is what's bugging me.

**Comment by galpeter** [comment]
lgtm

**Comment by zherczeg** [comment]
I think we can improve error messages in a follow-up patch, this patch focuses on removing the completion value and already big enough.

**Pull request merged**